students/elaine_x/session08/sparse_array_ex.py

```diff
@@ -0,0 +1,127 @@
+'''
+##########################
+#Python 210
+#Session 08 - Sparse Array
+#Elaine Xu
+#Mar 8, 2019
+###########################
+'''
+'''
+https://uwpce-pythoncert.github.io/PythonCertDevel/exercises/sparse_array.html
+
+Oftentimes, at least in computation programming, we have large arrays of data that hold mostly zeros
+These are referred to as “sparse” as the information in them is widely scattered, or sparse.
+Since they are mostly zeros, it can be memory and computationally efficient to store only the value
+that are non-zero.
+But you want it to look like a regular array in user code.
+In the real world, these are usually 2 dimensional arrays. But to keep it a bit simpler, we’ll make
+a 1 dimensional sparse array in this class.
+
+A sparse array class should present to the user the same interface as a regular list.
+
+Internally, it can store the values in a dict, with the index as the keys, so that only the indexes
+with non-zero values will be stored.
+It should take a sequence of values as an initializer:
+sa = SparseArray([1,2,0,0,0,0,3,0,0,4])
+
+you should be able to tell how long it is:
+len(my_array)
+This will give its “virtual” length – with the zeros
+
+It should support getting and setting particular elements via indexing:
+sa[5] = 12
+sa[3] = 0 # the zero won't get stored!
+val = sa[13] # it should get a zero if not set
+
+It should support deleting an element by index:
+del sa[4]
+
+It should raise an IndexError if you try to access an index beyond the end.
+it should have an append() method.
+Can you make it support slicing?
+How else can you make it like a list?
+'''
+class SparseArray():
+    '''generates sparse array'''
+    def __init__(self, array):
+        self.array = {}
+        self.len = len(array)
+        for i, item in enumerate(array):
+            if item is not 0:
+                self.array[i] = item
+
+    def __str__(self):
+        return "Sparse Array({})".format(self.array)
```

Comment:
This one should have a return that looks more like a list, as the sparse array acts as compression for lists with lots of zero values. You could also consider implementing the _repr_ method and have it return a list.